Scripts/Editor/AssetImport/PreventAutoRefreshWhilePlaying.cs

```diff
@@ -0,0 +1,108 @@
+using Anvil.CSharp.Logging;
+using UnityEditor;
+
+namespace Anvil.Unity.Editor.AssetImport
+{
+    /// <summary>
+    /// Prevents the asset database from refreshing while the editor is in play mode.
+    /// This means that any changed assets aren't imported and changed source isn't recompiled.
+    ///
+    /// NOTE: When this is enabled it overrides the behaviour of
+    /// Preferences -> General -> Script Changes While Playing
+    /// (In versions of Unity that include that option)
+    /// </summary>
+    [InitializeOnLoad]
+    public static class PreventAutoRefreshWhilePlaying
+    {
+        private const string MENU_PATH = "Anvil/Prevent Asset Refresh While Playing";
+        private const string PREFSKEY_IS_ENABLED = MENU_PATH + ":IsEnabled";
+
+        private static readonly Logger s_Logger = Log.GetStaticLogger(typeof(PreventAutoRefreshWhilePlaying));
+        private static bool s_IsAutoRefreshDisabled = false;
+
+        private static bool IsEnabled
+        {
+            get => EditorPrefs.GetBool(PREFSKEY_IS_ENABLED);
+            set => EditorPrefs.SetBool(PREFSKEY_IS_ENABLED, value);
+        }
+
+        static PreventAutoRefreshWhilePlaying()
+        {
+            EditorApplication.playModeStateChanged += EditorApplication_PlayModeStateChanged;
+        }
+
+        [MenuItem(MENU_PATH)]
+        private static void ToggleEnabled()
+        {
+            IsEnabled = !IsEnabled;
+
+            if (IsEnabled)
+            {
+                EditorUtility.DisplayDialog(
+                    "Note",
+                    "Enabling this option prevents assets and scripts from reloading while the editor is playing. This overrides the behaviour of \"Preferences -> General -> Script Changes While Playing\".",
+                    "Thanks Boss"
+                    );
+            }
+
+            bool shouldPreventRefresh = IsEnabled && EditorApplication.isPlaying;
+            if (shouldPreventRefresh)
+            {
+                PreventAutoRefresh();
+            }
+            else
+            {
+                AllowAutoRefresh();
+            }
+        }
+
+        [MenuItem(MENU_PATH, true)]
+        private static bool ToggleEnabled_Validator()
+        {
+            // Make sure the menu is showing the correct checked state
+            Menu.SetChecked(MENU_PATH, IsEnabled);
+            return true;
+        }
+
+        private static void EditorApplication_PlayModeStateChanged(PlayModeStateChange state)
```

Comment:
Does this get called multiple times with the domain reloading turned off? Not sure if it affects Editor static events the same way or not but we should check and guard against it if it does.

Reply:
Not sure the concern you're worried about.

If the concern is that this gets called multiple times for a `state` during a single transition from play to edit (or vice versa) then it's fine. The `Allow/PreventAutoRefresh()` methods are well protected against being called multiple times. They have to be because `AssetDatabase.DisallowAutoRefresh();` and `AssetDatabase.AllowAutoRefresh();` are a bit dangerous. They increment and decrement a private counter so if you get these calls out of balance you can really break the state of the editor.

If the concern is that this doesn't called when domain reload is turned off, it does. The logs emitted by `Allow/PreventAutoRefresh()` are shown on every transition.

Reply:
My concern was that the static constructor would get called multiple times during development after every recompile but there wasn't ever a removal from that listener. 

Spend a few days without closing Unity and you have a build up of the function getting hammered by multiple listeners. 

I know you've gated the heavy work which is great, I was just concerned about a hidden runaway slow down. 

I did a quick test and it doesn't seem to be the case. Logs in the function only trigger once so we're good. 

You're free to merge.